File: R/optindval.R
optindval <- function (veg,clustering,maxitr=100,minsiz=5) 
{
    numplt <- nrow(veg)
    numspc <- ncol(veg)
    if (inherits(clustering, c("partana", "clustering", "partition"))) 
        clustering = clustering$clustering
    if (is.numeric(clustering)) {
        if (min(clustering)< 0 || (length(table(clustering)) != max(clustering))) {
            cat('WARNING: renumbering clusters to consecutive integers\n')
            clustering <- match(clustering,sort(unique(clustering)))
        }
    }

    numcls <- max(clustering)
    relfrq <- matrix(0,nrow=numspc,ncol=numcls)
    relabu <- matrix(0,nrow=numspc,ncol=numcls)
    indval <- matrix(0,nrow=numspc,ncol=numcls)
    pval <- rep(0,numspc)
    indcls <- rep(0,numspc)
    clstab <- rep(0,numcls)
    maxcls <- rep(0,numspc)
    sumind <- 0
    sums <- rep(0,maxitr+1)
    numitr <- 0
    tmpclu <- rep(0,numplt)
    res <- .Fortran('optindval',
                   as.double(as.matrix(veg)),
                   as.integer(numplt),
                   as.integer(numspc),
                   clustering = as.integer(clustering),
                   as.integer(numcls),
                   as.double(relfrq),
                   as.double(relabu),
                   as.double(indval),
                   as.double(indcls),
                   as.integer(clstab),
                   as.integer(maxcls),
                   as.double(sumind),
                   as.integer(maxitr),
                   as.integer(minsiz),
                   sums = as.double(sums),
                   numitr = as.integer(numitr),
                   as.integer(tmpclu),
                   PACKAGE='optpart')
    out <- indval(veg,res$clustering)
    out$clustering <- res$clustering
    out$sums <- res$sums[1:res$numitr]
    out$numitr <- res$numitr
    class(out) <- c('optindval','clustering')
    return(out)
}

File: src/optindval.f90
      subroutine optindval(veg,numplt,numspc,clusid,numcls,  &
                        relfrq,relabu,indval,indcls,clstab,  &
                        maxcls,sumind,maxitr,minsiz,sums,    &
                        numitr,tmpclu)
 
!* passed in/out
 
      integer numplt
      integer numspc
      integer clusid(numplt)
      integer numcls
      integer maxitr
      integer minsiz
      integer numitr
      double precision veg(numplt,numspc)
      double precision sums(0:maxitr)
 
!* passed through to clcdul
 
      double precision relfrq(numspc,numcls)
      double precision relabu(numspc,numcls)
      double precision indval(numspc,numcls)
      double precision indcls(numspc)
      double precision sumind
      integer clstab(numcls)
      integer maxcls(numspc)
 
!* scratch arrays
 
      integer tmpclu(numplt)

!* local
 
      double precision base
      integer moveto
      integer mvefrm
      integer plttrn
      integer flag
 
!********************* one ******************************
 
      call clcdul(veg,numplt,numspc,clusid,numcls,relfrq,relabu,   &
                  indval,indcls,clstab,maxcls,sumind)

      base = sumind
      numitr = maxitr
      sums(0) = sumind

      do i=1,numcls
        clstab(i) = 0
      end do    

      do i=1,numplt
        clstab(clusid(i)) = clstab(clusid(i)) + 1
      end do   
 
      mvefrm = 0
      moveto = 0
      plttrn = 0

      do k=1,maxitr
        flag = 0
        do i=1,numplt
          mvefrm = 0
          moveto = 0
          plttrn = 0
          do j=1,numplt
            tmpclu(j) = clusid(j)
          end do  
 
          do j=1,numcls
            if (j .eq. clusid(i)) cycle
            if (clstab(clusid(i)) .le. minsiz) cycle
            tmpclu(i) = j
            call clcdul(veg,numplt,numspc,tmpclu,numcls,relfrq,     &
                        relabu,indval,indcls,clstab,maxcls,sumind)
            if (sumind .gt. base) then
              base = sumind
              moveto = tmpclu(i)
              mvefrm = clusid(i)
              plttrn = i
              flag = 1
            endif
          end do   
        end do  
 
        if (flag .eq. 1) then
          clusid(plttrn) = moveto
          clstab(moveto) = clstab(moveto) + 1
          clstab(mvefrm) = clstab(mvefrm) - 1
          sums(k) = base
          sumind = base
        else
          numitr = k 
          return
        endif
      end do   
 
      return
 
      end
